src/sage/rings/polynomial/integer_valued_polynomials.py

```diff
@@ -0,0 +1,1125 @@
+    `\binom{x}{n}` for `n \geq 0` (the *binomial basis*) and of
+    the other sequence `\binom{x+n}{n}` for `n \geq 0` (the *shifted basis*).
+
+    TESTS::
+
+        sage: IntegerValuedPolynomialRing(24)
+        Traceback (most recent call last):
+        ...
+        TypeError: argument R must be a ring
+    """
+    def __init__(self, R):
+        if R not in Rings():
+            raise TypeError("argument R must be a ring")
+        self._base = R
+        cat = Algebras(R).Commutative().WithBasis()
+        Parent.__init__(self, base=R, category=cat.WithRealizations())
+
+        S = self.S()
+        B = self.B()
+        B.module_morphism(S._from_binomial_basis, codomain=S).register_as_coercion()
+        S.module_morphism(B._from_shifted_basis, codomain=B).register_as_coercion()
+
+    _shorthands = ["B", "S"]
+
+    def a_realization(self):
+        r"""
+        Return the default realization of ``self``.
+
+        This is the shifted basis.
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.a_realization()
+            Integer-Valued Polynomial Ring over Rational Field
+            in the shifted basis
+        """
+        return self.S()
+
+    def indices(self):
+        r"""
+        Return the objects that index the basis elements of this algebra.
+
+        This is the set `\NN` of nonnegative integers.
+
+        EXAMPLES::
+
+            sage: A = IntegerValuedPolynomialRing(QQ)
+            sage: A.indices()
+            Non negative integers
+        """
+        return NonNegativeIntegers()
+
+    def _repr_(self) -> str:
+        r"""
+        Return the string representation.
+
+        EXAMPLES::
+
+            sage: IntegerValuedPolynomialRing(QQ)
+            Integer-Valued Polynomial Ring over Rational Field
+        """
+        br = self.base_ring()
+        return f"Integer-Valued Polynomial Ring over {br}"
+
+    class Bases(Category_realization_of_parent):
+        def super_categories(self) -> list:
+            r"""
+            Return the super-categories of ``self``.
+
+            EXAMPLES::
+
+                sage: A = IntegerValuedPolynomialRing(QQ); A
+                Integer-Valued Polynomial Ring over Rational Field
+                sage: C = A.Bases(); C
+                Category of bases of Integer-Valued Polynomial Ring
+                over Rational Field
+                sage: C.super_categories()
+                [Category of realizations of Integer-Valued Polynomial Ring
+                 over Rational Field,
+                 Join of Category of algebras with basis over Rational Field and
+                 Category of commutative algebras over Rational Field and
+                 Category of realizations of unital magmas]
+            """
+            A = self.base()
+            category = Algebras(A.base_ring()).Commutative()
+            return [A.Realizations(),
+                    category.Realizations().WithBasis()]
```

Comment:
```suggestion
            category = Algebras(A.base_ring()).Commutative()
            return [A.Realizations(),
                    category.Realizations().WithBasis(),
                    category.WithBasis()]
```
Isn't there also a natural filtered algebra structure?

Reply:
indeed, done (hopefully)